torch/csrc/autograd/engine.cpp

```diff
@@ -225,21 +226,34 @@ Engine::Engine() : max_recursion_depth_(MAX_DEPTH), non_reentrant_device_thread_
 // Send shutdown tasks to all device_ready_queues_ if no backward tasks are running
 // Even though readyQueue should be empty, shutdown tasks have the highest priority
 Engine::~Engine() {
+  // Under some conditions, autograd threads can hang on shutdown
+  // Do not wait for them to shutdown indefinitely but rely on timeout
+  auto wait_duration_str = getenv("TORCH_AUTOGRAD_SHUTDOWN_WAIT_LIMIT");
+  if (!wait_duration_str) {
+    wait_duration_str = "10.0";
+  }
+  auto wait_duration = std::atof(wait_duration_str);
```

Comment:
One little hitch here: if `TORCH_AUTOGRAD_SHUTDOWN_WAIT_LIMIT` is not specified, we default to 10s, but if it is specified and invalid (atof doesn't understand it), we default to 0s. I guess there isn't really too much you can do about it because atof doesn't error report.